drivers/clock_control/clock_stm32_ll_h7.c

```diff
@@ -31,6 +31,97 @@
+#ifdef CONFIG_CPU_CORTEX_M7
+
+/* Define primary oscillator frequencies */
+/* Suppress the cast to uint32_t which */
+/* prevents from compare with #if > operator*/
+/* original defines in stm32h7xx_hal_conf.h*/
+#define HSI_FREQ 64000000UL	/* HSI_VALUE ((uint32_t)64000000) */
+/* HSE_VALUE overridden by the build system without C cast to uint*/
+/* Build system doesn't provide the UL type suffix for HSE_VALUE */
+/* Force HSE_FREQ to be cast to preprocessor UL to prevent overflow*/
+/* Only use these constants in preprocessor expressions */
+#define HSE_FREQ HSE_VALUE
+#define CSI_FREQ 4000000UL	/* CSI_VALUE ((uint32_t)4000000) */
+
+/* Choose PLL SRC */
+#ifdef CONFIG_CLOCK_STM32_PLL_SRC_HSI
+#define PLLSRC_FREQ  HSI_FREQ
+#elif defined(CONFIG_CLOCK_STM32_PLL_SRC_CSI)
+#define PLLSRC_FREQ  CSI_FREQ
+#elif defined(CONFIG_CLOCK_STM32_PLL_SRC_HSE)
+#define PLLSRC_FREQ  HSE_FREQ
+#else
+#define PLLSRC_FREQ 0
+#endif
+
+/* Given source clock and dividers, computed the output frequency of PLLP */
+#define PLLP_FREQ(pllsrc_freq, divm, divn, divp)	(((pllsrc_freq)*\
+							(divn))/((divm)*(divp)))
+
+/* PLL P output frequency value */
+#define PLLP_VALUE	PLLP_FREQ(\
+				PLLSRC_FREQ,\
+				CONFIG_CLOCK_STM32_PLL_M_DIVISOR,\
+				CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER,\
+				CONFIG_CLOCK_STM32_PLL_P_DIVISOR)
+
+/* SYSCLKSRC before the D1CPRE prescaler */
+#ifdef CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL
+#define SYSCLKSRC_FREQ	PLLP_VALUE
+#elif defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI)
+#define SYSCLKSRC_FREQ	HSI_FREQ
+#elif defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_CSI)
+#define SYSCLKSRC_FREQ	CSI_FREQ
+#elif defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_HSE)
+#define SYSCLKSRC_FREQ	HSE_FREQ
+#endif
+
+/* ARM Sys CPU Clock before HPRE prescaler */
+#define SYSCLK_FREQ	((SYSCLKSRC_FREQ)/(CONFIG_CLOCK_STM32_D1CPRE))
+#define AHB_FREQ	((SYSCLK_FREQ)/(CONFIG_CLOCK_STM32_HPRE))
+#define APB1_FREQ	((AHB_FREQ)/(CONFIG_CLOCK_STM32_D2PPRE1))
+#define APB2_FREQ	((AHB_FREQ)/(CONFIG_CLOCK_STM32_D2PPRE2))
+#define APB3_FREQ	((AHB_FREQ)/(CONFIG_CLOCK_STM32_D1PPRE))
+#define APB4_FREQ	((AHB_FREQ)/(CONFIG_CLOCK_STM32_D3PPRE))
+
+/* Datasheet maximum frequency definitions */
+#define SYSCLK_FREQ_MAX	480000000UL
+#define AHB_FREQ_MAX	240000000UL
+#define APBx_FREQ_MAX	120000000UL
+
+#if SYSCLK_FREQ > SYSCLK_FREQ_MAX
+#error "SYSCLK frequency is too high, max is 480MHz"
+#endif
+#if AHB_FREQ > AHB_FREQ_MAX
+#error "AHB frequency is too high, max is 240MHz"
+#endif
+#if APB1_FREQ > APBx_FREQ_MAX
+#error "APB1 frequency is too high, max is 120MHz"
+#endif
+#if APB2_FREQ > APBx_FREQ_MAX
+#error "APB2 frequency is too high, max is 120MHz"
+#endif
+#if APB3_FREQ > APBx_FREQ_MAX
+#error "APB3 frequency is too high, max is 120MHz"
+#endif
+#if APB4_FREQ > APBx_FREQ_MAX
+#error "APB4 frequency is too high, max is 120MHz"
+#endif
+
+#if SYSCLK_FREQ != CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
```

Comment:
Another staff:

Putting clock settings from below into M4 defconfig does not trigger this error, if `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` will be changed to any arbitrary value.

the settings:
```
# CSI PLL sysclk 240MHz
# CONFIG_CLOCK_STM32_PLL_SRC_HSE=y
CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
# CONFIG_CLOCK_STM32_PLL_SRC_CSI=y

CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y

# CSI n; HSI: n; HSE: y;
# CONFIG_CLOCK_STM32_HSE_BYPASS=n
CONFIG_CLOCK_STM32_HSE_CLOCK=8000000

CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=240000099

# CSI: ?; HSI: 8; HSE: x;
CONFIG_CLOCK_STM32_HSI_DIVISOR=8
# CSI: 4; HSI: 1; HSE 2;
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=1
# CSI: 240; HSI: 60; HSE ;
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=60
# CSI: 2; HSI: 1; HSE 2;
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2

CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```

Reply:
@Nukersson ,
On dual core (H745 , H743) the clock tree is common to both cores. So to avoid conflict, it has been decided on zephyr that only cortex M7 will perform Clock configuration (look at `#if !defined(CONFIG_CPU_CORTEX_M4)` in `stm32_clock_control_init()`).
So M4 will not configure clocks, this is why no check is performed on M4.
But the check will be done by M7, which ensure all frequencies (for both core) are in the proper range.
The point is that cortex M4 zephyr build still need to know the value of CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC .
so there is a constraint: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC  in M4 defconfig MUST match with M7 HCLK computation (computed from M7 defconfig).
But there is no mean to ensure such coherence between 2 boards within zephyr build system.

Reply:
Okay. Understand it.

Reply:
@ABOSTM is right.
Zephyr doesn't provide this possiblity and I have added a comment before the preprocessor check on why the clock tree check is only performed by the M7 core.

@Nukersson I cannot had the check for clock coherence in M4 core a @ABOSTM said. I tried trust me 😫 
The only thing that could be done is a runtime check but I don't think this would improve the situation much.

At least, the user will be prompt if its M7 clock config are not feasable.

But your idea for Pragma messages is great and could help the user in case of error detection 👍 I'm gonna test that right now 😄

Reply:
Gives this kind of outputs:
```
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:124:9: note: #pragma message: SYSCLK_FREQ is: (((((((64000000UL)/(4)))* (120))/((1)*(2))))/(1))
  124 | #pragma message "SYSCLK_FREQ is: " XSTR(SYSCLK_FREQ)
      |         ^~~~~~~
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:125:9: note: #pragma message: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: 480000000
  125 | #pragma message "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: "\
      |         ^~~~~~~
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:127:2: error: #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
  127 | #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
      |  ^~~~~
```
Do you still think it can help the user ? Maybe some message indicating to review prescalers/dividers to match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC whould be better.

Reply:
> Gives this kind of outputs:
> ```
> /home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:124:9: note: #pragma message: SYSCLK_FREQ is: (((((((64000000UL)/(4)))* (120))/((1)*(2))))/(1))
>   124 | #pragma message "SYSCLK_FREQ is: " XSTR(SYSCLK_FREQ)
>       |         ^~~~~~~
> /home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:125:9: note: #pragma message: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: 480000000
>   125 | #pragma message "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: "\
>       |         ^~~~~~~
> /home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:127:2: error: #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
>   127 | #error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
>       |  ^~~~~
> ```
> Do you still think it can help the user ? Maybe some message indicating to review prescalers/dividers to match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC whould be better.

I think `SYSCLK_FREQ` output is good. It is because of fact, that GCC preprocessor stringificiation does not evaluate the value, but just prints the define you proposed. Neverless, the user can see here, that one of the dividers/miltipliers (4, 120, 1, 2, 1) is wrong and how `SYSCLK_FREQ` will be built. This can give the user input, to adapt the dividers/miltipliers better.

Reply:
Alright, I'd like to know what others think about this but my suggestion is:

```c
#if SYSCLK_FREQ != CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
#error "Calculated CPU clock frequency (SYSCLK) for M7 core doesn't match \
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC."
/* C preprocessor stringification macros */
#define XSTR(x) STR(x)
#define STR(x) #x
#pragma message "SYSCLK_FREQ with given prescalers is: "\
XSTR(SYSCLK_FREQ)
#pragma message "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: "\
XSTR(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC)
#endif
```

I've moved STR/XSTR macros in the specific error case such that in case of redefinition, the error message is printed before the redefined symbol error (my guess...).

Giving:
```
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:121:2: error: #error "Calculated CPU clock frequency (SYSCLK) for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC."
  121 | #error "Calculated CPU clock frequency (SYSCLK) for M7 core doesn't match \
      |  ^~~~~
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:126:9: note: #pragma message: SYSCLK_FREQ with given prescalers is: (((((((64000000UL)/(4)))* (120))/((1)*(2))))/(1))
  126 | #pragma message "SYSCLK_FREQ with given prescalers is: "\
      |         ^~~~~~~
/home/dev/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:128:9: note: #pragma message: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: 480000000
  128 | #pragma message "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is: "\
```

Reply:
My personal though is that the first basic error message was good "Calculated CPU clock frequency (SYSCLK) for M7 core doesn't match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC."
With this error, user is able to check by himself why `SYSCLK_FREQ != CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC`
and revert computation.
So for me it is a bit too much. But as it is already done (and well done), I won't reject it. 
(By the way it a clever way to pass define value in preprocessor error)